src/db/mysql/models.rs

```diff
@@ -0,0 +1,352 @@
+use std::{self, ops::Deref};
+
+use diesel::{
+    insert_into,
+    mysql::MysqlConnection,
+    r2d2::{ConnectionManager, Pool, PooledConnection},
+    sql_types::{BigInt, Integer, Text},
+    update, Connection, ExpressionMethods, OptionalExtension,
+};
+use diesel::{sql_query, QueryDsl, RunQueryDsl};
+#[cfg(test)]
+use diesel_logger::LoggingConnection;
+use futures::future;
+
+use super::schema::{bso, collections, user_collections};
+#[cfg(test)]
+use super::test::TestTransactionCustomizer;
+use db::{
+    error::DbError, get_std_collection_id, get_std_collection_name, params, results,
+    util::ms_since_epoch, Db, DbFuture, Sorting,
+};
+use settings::Settings;
+
+embed_migrations!();
+
+pub type Result<T> = std::result::Result<T, DbError>;
+
+// The ttl to use for rows that are never supposed to expire (in seconds)
+pub const DEFAULT_BSO_TTL: u32 = 2100000000;
+
+no_arg_sql_function!(last_insert_id, Integer);
+
+/// Run the diesel embedded migrations
+///
+/// Mysql DDL statements implicitly commit which could disrupt MysqlPool's
+/// begin_test_transaction during tests. So this runs on its own separate conn.
+pub fn run_embedded_migrations(settings: &Settings) -> Result<()> {
+    let conn = MysqlConnection::establish(&settings.database_url).unwrap();
+    Ok(embedded_migrations::run(&conn)?)
+}
+
+pub struct MysqlDbPool {
+    pool: Pool<ConnectionManager<MysqlConnection>>,
+}
+
+// XXX: to become a db::DbPool trait
+impl MysqlDbPool {
+    pub fn new(settings: &Settings) -> Result<Self> {
+        let manager = ConnectionManager::<MysqlConnection>::new(settings.database_url.as_ref());
+        let mut builder = Pool::builder().max_size(settings.database_pool_max_size.unwrap_or(10));
+        #[cfg(test)]
+        {
+            if settings.database_use_test_transactions {
+                builder = builder.connection_customizer(Box::new(TestTransactionCustomizer));
+            }
+        }
+        Ok(Self {
+            pool: builder.build(manager)?,
+        })
+    }
+
+    pub fn get(&self) -> Result<MysqlDb> {
+        Ok(MysqlDb {
+            #[cfg(not(test))]
+            conn: self.pool.get()?,
+            #[cfg(test)]
+            conn: LoggingConnection::new(self.pool.get()?),
+        })
+    }
+}
+
+pub struct MysqlDb {
+    #[cfg(not(test))]
+    pub(super) conn: PooledConnection<ConnectionManager<MysqlConnection>>,
+    #[cfg(test)]
+    pub(super) conn: LoggingConnection<PooledConnection<ConnectionManager<MysqlConnection>>>,
+}
+
+impl MysqlDb {
+    pub fn get_collection_id_sync(
```

Comment:
It's not obvious to me why some of these are suffixed with `_sync` and some are not; what's the significance?

Reply:
the higher level db interface (trait) supplies async calls. So the MysqlDb impl of this trait will call sync methods via the [tokio blocking](https://docs.rs/tokio-threadpool/0.1/tokio_threadpool/fn.blocking.html) wrapping call.